File: multi-processor-programming-study/solutions/cpp/work_stealing_deque_solution.cpp
/**
 * SOLUTION: Work-Stealing Deque (Chase-Lev Algorithm)
 *
 * This is the complete reference implementation of the Chase-Lev work-stealing deque.
 * All TODOs from the exercise have been completed.
 *
 * KEY IMPLEMENTATION POINTS:
 * 1. Owner operations (push/pop) use relaxed memory ordering when possible
 * 2. Thief operations (steal) use acquire/release and CAS
 * 3. Dynamic resizing handled by owner only
 * 4. Memory fences ensure correct synchronization
 */

#include <iostream>
#include <atomic>
#include <thread>
#include <vector>
#include <memory>
#include <chrono>
#include <random>
#include <algorithm>

namespace multiprocessor::part2::work_stealing::solution {

template <typename T>
class WorkStealingDeque {
private:
    struct CircularArray {
        std::unique_ptr<std::atomic<T>[]> buffer;
        size_t capacity;

        explicit CircularArray(size_t size) : capacity(size) {
            buffer = std::make_unique<std::atomic<T>[]>(capacity);
        }

        T get(size_t index) const {
            return buffer[index & (capacity - 1)].load(std::memory_order_relaxed);
        }

        void put(size_t index, T value) {
            buffer[index & (capacity - 1)].store(value, std::memory_order_relaxed);
        }

        std::unique_ptr<CircularArray> grow(size_t bottom, size_t top) {
            auto new_array = std::make_unique<CircularArray>(capacity * 2);
            for (size_t i = top; i < bottom; ++i) {
                new_array->put(i, get(i));
            }
            return new_array;
        }
    };

    std::atomic<size_t> top_{0};
    std::atomic<size_t> bottom_{0};
    std::atomic<CircularArray*> array_;

    static constexpr size_t INITIAL_CAPACITY = 256;

public:
    WorkStealingDeque() {
        CircularArray* initial_array = new CircularArray(INITIAL_CAPACITY);
        array_.store(initial_array, std::memory_order_relaxed);
    }

    ~WorkStealingDeque() {
        delete array_.load(std::memory_order_relaxed);
    }

    /**
     * SOLUTION: push (owner only)
     */
    void push(T value) {
        size_t b = bottom_.load(std::memory_order_relaxed);
        size_t t = top_.load(std::memory_order_acquire);
        CircularArray* a = array_.load(std::memory_order_relaxed);

        if (b - t >= a->capacity) {
            CircularArray* new_array = a->grow(b, t).release();
            array_.store(new_array, std::memory_order_release);
            delete a;
            a = new_array;
        }

        a->put(b, value);
        std::atomic_thread_fence(std::memory_order_release);
        bottom_.store(b + 1, std::memory_order_relaxed);
    }

    /**
     * SOLUTION: pop (owner only)
     */
    T pop() {
        size_t b = bottom_.load(std::memory_order_relaxed) - 1;
        CircularArray* a = array_.load(std::memory_order_relaxed);
        bottom_.store(b, std::memory_order_relaxed);

        std::atomic_thread_fence(std::memory_order_seq_cst);

        size_t t = top_.load(std::memory_order_relaxed);

        T result{};
        if (t <= b) {
            result = a->get(b);

            if (t == b) {
                if (!top_.compare_exchange_strong(t, t + 1,
                                                  std::memory_order_seq_cst,
                                                  std::memory_order_relaxed)) {
                    result = T{};
                }
                bottom_.store(b + 1, std::memory_order_relaxed);
            }
        } else {
            bottom_.store(b + 1, std::memory_order_relaxed);
        }

        return result;
    }

    /**
     * SOLUTION: steal (thieves)
     */
    T steal() {
        size_t t = top_.load(std::memory_order_acquire);
        std::atomic_thread_fence(std::memory_order_seq_cst);
        size_t b = bottom_.load(std::memory_order_acquire);

        if (t >= b) {
            return T{};
        }

        CircularArray* a = array_.load(std::memory_order_consume);
        T value = a->get(t);

        if (!top_.compare_exchange_strong(t, t + 1,
                                         std::memory_order_seq_cst,
                                         std::memory_order_relaxed)) {
            return T{};
        }

        return value;
    }

    size_t size() const {
        size_t b = bottom_.load(std::memory_order_relaxed);
        size_t t = top_.load(std::memory_order_relaxed);
        return (b >= t) ? (b - t) : 0;
    }
};

// Test task
struct Task {
    int id;
    int workload;

    Task() : id(0), workload(0) {}
    Task(int i, int w) : id(i), workload(w) {}

    void execute() const {
        volatile int sum = 0;
        for (int i = 0; i < workload; ++i) {
            sum += i;
        }
    }
};

void demonstrate_solution() {
    std::cout << "=== SOLUTION: Work-Stealing Deque ===\n\n";

    std::cout << "Implementation Details:\n\n";

    std::cout << "1. push() - Owner pushes to bottom:\n";
    std::cout << "   - Load bottom with relaxed ordering\n";
    std::cout << "   - Load top with acquire (sync with thieves)\n";
    std::cout << "   - Check if resize needed, grow if necessary\n";
    std::cout << "   - Put value at bottom\n";
    std::cout << "   - Release fence + store bottom\n\n";

    std::cout << "2. pop() - Owner pops from bottom:\n";
    std::cout << "   - Decrement bottom (claim task)\n";
    std::cout << "   - Sequential fence (sync with steal)\n";
    std::cout << "   - Load top\n";
    std::cout << "   - If last element, CAS with thieves\n";
    std::cout << "   - Otherwise just take it\n\n";

    std::cout << "3. steal() - Thieves steal from top:\n";
    std::cout << "   - Load top with acquire\n";
    std::cout << "   - Sequential fence\n";
    std::cout << "   - Load bottom with acquire\n";
    std::cout << "   - Check if empty\n";
    std::cout << "   - Read value\n";
    std::cout << "   - CAS top to claim\n\n";

    // Demonstration
    WorkStealingDeque<Task> deque;

    std::cout << "Test: Basic operations\n";
    deque.push(Task(1, 100));
    deque.push(Task(2, 200));
    deque.push(Task(3, 300));

    std::cout << "Pushed tasks: 1, 2, 3\n";
    std::cout << "Deque size: " << deque.size() << "\n\n";

    Task t1 = deque.pop();
    std::cout << "Popped from owner: " << t1.id << " (LIFO - should be 3)\n";

    Task t2 = deque.steal();
    std::cout << "Stolen by thief: " << t2.id << " (FIFO - should be 1)\n\n";

    std::cout << "✅ Solution demonstrates correct Chase-Lev algorithm\n";
    std::cout << "✅ Owner gets LIFO order (cache locality)\n";
    std::cout << "✅ Thieves get FIFO order (load balancing)\n";
}

} // namespace

int main() {
    multiprocessor::part2::work_stealing::solution::demonstrate_solution();
    return 0;
}

File: multi-processor-programming-study/solutions/java/WorkStealingDeque_Solution.java
/**
 * SOLUTION: Work-Stealing Deque (Chase-Lev Algorithm)
 *
 * This is the complete reference implementation of the Chase-Lev work-stealing deque in Java.
 * All TODOs from the exercise have been completed.
 *
 * KEY IMPLEMENTATION POINTS:
 * 1. Owner operations (push/pop) are fast and mostly non-blocking
 * 2. Thief operations (steal) use CAS and may fail
 * 3. Dynamic resizing handled by owner only
 * 4. Similar to what Java's ForkJoinPool uses internally
 */

package solutions;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import java.util.concurrent.atomic.AtomicReferenceArray;

public class WorkStealingDeque_Solution<T> {

    /**
     * Circular array for storing tasks
     */
    private static class CircularArray<T> {
        private final AtomicReferenceArray<T> buffer;
        private final int capacity;

        CircularArray(int size) {
            this.capacity = size;
            this.buffer = new AtomicReferenceArray<>(size);
        }

        T get(long index) {
            return buffer.get((int) (index & (capacity - 1)));
        }

        void put(long index, T value) {
            buffer.set((int) (index & (capacity - 1)), value);
        }

        CircularArray<T> grow(long bottom, long top) {
            CircularArray<T> newArray = new CircularArray<>(capacity * 2);
            for (long i = top; i < bottom; i++) {
                newArray.put(i, get(i));
            }
            return newArray;
        }

        int getCapacity() {
            return capacity;
        }
    }

    private final AtomicInteger top = new AtomicInteger(0);
    private final AtomicInteger bottom = new AtomicInteger(0);
    private final AtomicReference<CircularArray<T>> array;

    private static final int INITIAL_CAPACITY = 256;

    public WorkStealingDeque_Solution() {
        array = new AtomicReference<>(new CircularArray<>(INITIAL_CAPACITY));
    }

    /**
     * SOLUTION: push (owner only)
     *
     * Fast path - no CAS needed, just atomic stores
     */
    public void push(T value) {
        long b = bottom.get();
        long t = top.get();
        CircularArray<T> a = array.get();

        // Resize if needed
        if (b - t >= a.getCapacity()) {
            CircularArray<T> newArray = a.grow(b, t);
            array.set(newArray);
            a = newArray;
        }

        // Put value and increment bottom
        a.put(b, value);
        bottom.set((int) (b + 1));
    }

    /**
     * SOLUTION: pop (owner only)
     *
     * Fast path when no competition, CAS only for last element
     */
    public T pop() {
        long b = bottom.get() - 1;
        CircularArray<T> a = array.get();
        bottom.set((int) b);

        long t = top.get();

        T result = null;
        if (t <= b) {
            // Non-empty
            result = a.get(b);

            if (t == b) {
                // Last element - compete with thieves
                if (!top.compareAndSet((int) t, (int) (t + 1))) {
                    // Lost to thief
                    result = null;
                }
                bottom.set((int) (b + 1));
            }
        } else {
            // Empty
            bottom.set((int) (b + 1));
        }

        return result;
    }

    /**
     * SOLUTION: steal (thieves)
     *
     * Always uses CAS, can fail if racing with other thieves or owner
     */
    public T steal() {
        long t = top.get();
        long b = bottom.get();

        if (t >= b) {
            return null; // Empty
        }

        CircularArray<T> a = array.get();
        T value = a.get(t);

        // Try to claim with CAS
        if (!top.compareAndSet((int) t, (int) (t + 1))) {
            return null; // Lost race
        }

        return value;
    }

    public int size() {
        int b = bottom.get();
        int t = top.get();
        return Math.max(0, b - t);
    }

    /**
     * Demonstration of the solution
     */
    public static void main(String[] args) {
        System.out.println("=== SOLUTION: Work-Stealing Deque ===\n");

        System.out.println("Implementation Details:\n");

        System.out.println("1. push() - Owner pushes to bottom:");
        System.out.println("   - Get current bottom and top");
        System.out.println("   - Check if resize needed");
        System.out.println("   - Put value at bottom index");
        System.out.println("   - Increment bottom (atomic store)\n");

        System.out.println("2. pop() - Owner pops from bottom:");
        System.out.println("   - Decrement bottom (claim task)");
        System.out.println("   - Get value at bottom");
        System.out.println("   - If last element, CAS with thieves");
        System.out.println("   - Otherwise just return it\n");

        System.out.println("3. steal() - Thieves steal from top:");
        System.out.println("   - Get top and bottom");
        System.out.println("   - Check if empty");
        System.out.println("   - Read value at top");
        System.out.println("   - CAS top to claim it\n");

        // Test
        WorkStealingDeque_Solution<Integer> deque = new WorkStealingDeque_Solution<>();

        System.out.println("Test: Basic operations");
        deque.push(1);
        deque.push(2);
        deque.push(3);

        System.out.println("Pushed: 1, 2, 3");
        System.out.println("Size: " + deque.size());

        Integer owner = deque.pop();
        System.out.println("Owner popped: " + owner + " (LIFO - should be 3)");

        Integer thief = deque.steal();
        System.out.println("Thief stole: " + thief + " (FIFO - should be 1)\n");

        System.out.println("✅ Solution implements Chase-Lev correctly");
        System.out.println("✅ Owner: LIFO (cache locality)");
        System.out.println("✅ Thieves: FIFO (load balancing)");
        System.out.println("✅ Used in Java ForkJoinPool");
    }
}
